packages/gensx/src/checkpoint.ts

```diff
@@ -199,6 +220,217 @@ export class CheckpointManager implements CheckpointWriter {
+
+    // If either isn't an object, they're not equal
+    if (!a || !b || typeof a !== "object" || typeof b !== "object") {
+      return false;
+    }
+
+    // Handle arrays
+    if (Array.isArray(a) && Array.isArray(b)) {
+      return (
+        a.length === b.length &&
+        a.every((item, index) => this.isEqual(item, b[index]))
+      );
+    }
+
+    // Handle objects
+    if (!Array.isArray(a) && !Array.isArray(b)) {
+      const aKeys = Object.keys(a);
+      const bKeys = Object.keys(b);
+      return (
+        aKeys.length === bKeys.length &&
+        aKeys.every(key =>
+          this.isEqual(a[key as keyof typeof a], b[key as keyof typeof b]),
+        )
+      );
+    }
+
+    return false;
+  }
+
+  private withNode<T>(nodeId: string, fn: () => T): T {
+    this.currentNodeChain.push(nodeId);
+    try {
+      return fn();
+    } finally {
+      this.currentNodeChain.pop();
+    }
+  }
+
+  private getEffectiveSecrets(): Set<unknown> {
+    const allSecrets = new Set<unknown>();
+    // Collect secrets from current node and all ancestors
+    for (const nodeId of this.currentNodeChain) {
+      const nodeSecrets = this._secretValues.get(nodeId);
+      if (nodeSecrets) {
+        for (const secret of nodeSecrets) {
+          allSecrets.add(secret);
+        }
+      }
+    }
+    return allSecrets;
+  }
+
+  private registerSecrets(
+    props: Record<string, unknown>,
+    paths: string[],
+    nodeId: string,
+  ) {
+    this.withNode(nodeId, () => {
+      // Initialize secrets set for this node
+      let nodeSecrets = this._secretValues.get(nodeId);
+      if (!nodeSecrets) {
+        nodeSecrets = new Set();
+        this._secretValues.set(nodeId, nodeSecrets);
+      }
+
+      // Use paths purely for collection
+      for (const path of paths) {
+        const value = this.getValueAtPath(props, path);
+        if (value !== undefined) {
+          this.collectSecretValues(value, nodeSecrets);
+        }
+      }
+    });
+  }
+
+  private collectSecretValues(
+    data: unknown,
+    nodeSecrets: Set<unknown>,
+    visited = new WeakSet(),
+  ): void {
+    // Skip if already visited to prevent cycles
+    if (data && typeof data === "object") {
+      if (visited.has(data)) {
+        return;
+      }
+      visited.add(data);
+    }
+
+    // Handle primitive values
+    if (typeof data === "string") {
+      if (data.length >= this.MIN_SECRET_LENGTH) {
+        nodeSecrets.add(data);
+      }
+      return;
+    }
+
+    // Skip other primitives
+    if (!data || typeof data !== "object") {
+      return;
+    }
+
+    // Handle arrays and objects (excluding ArrayBuffer views)
+    if (Array.isArray(data) || !ArrayBuffer.isView(data)) {
+      const values = Array.isArray(data) ? data : Object.values(data);
+      values.forEach(value => {
+        this.collectSecretValues(value, nodeSecrets, visited);
+      });
+    }
+  }
+
+  private scrubSecrets(data: unknown, nodeId?: string, path = ""): unknown {
```

Comment:
Is `path` used for anything? I don't see it here.

Reply:
Yes, it is used for handling objects. 

```ts
      // Handle objects (excluding ArrayBuffer views)
      if (!ArrayBuffer.isView(data)) {
        return Object.fromEntries(
          Object.entries(data).map(([key, value]) => [
            key,
            this.scrubSecrets(value, nodeId, path ? `${path}.${key}` : key),
          ]),
        );
      }
```